>a) Construya un dataframe con los datos a analizar. Determine cuántas clases existen, cuántos registros por clase y describa el dataset.

In [383]:
import numpy as np
import pandas as pd
df = pd.read_csv('text_emotion.csv')

>b) Construya un conjunto de entrenamiento y otro de pruebas, a través de una máscara aleatoria, para verificar los resultados de los algoritmos. Genere un conjunto de validación si estima conveniente.

Este paso lo realizamos después de filtrar y preprocesar los datos

>c) Construya las representaciones de los datos con los que trabajará, ya sea para las entradas de los modelos como para las salidas. Recuerde que tendrá que codificar las distintas clases como valores numéricos enteros.

empty=0
sadness=1
enthusiasm=2
neutral=3
worry=4
surprise=5
love=6
fun=7
happines=8
hate=9
boredom=10
anger=11
relief=12
40000

In [384]:
print (df.sentiment.value_counts())

neutral       8638
worry         8459
happiness     5209
sadness       5165
love          3842
surprise      2187
fun           1776
relief        1526
hate          1323
empty          827
enthusiasm     759
boredom        179
anger          110
Name: sentiment, dtype: int64


In [385]:
sentimentint = np.empty(40000, dtype=object)
for i in range (40000):
    if df['sentiment'][i]=='empty':
        sentimentint[i]='0'
        continue
    if df['sentiment'][i]=='sadness':
        sentimentint[i]='1'
        continue
    if df['sentiment'][i]=='enthusiasm':
        sentimentint[i]='2'
        continue
    if df['sentiment'][i]=='neutral':
        sentimentint[i]='3'
        continue
    if df['sentiment'][i]=='worry':
        sentimentint[i]='4'
        continue
    if df['sentiment'][i]=='surprise':
        sentimentint[i]='5'
        continue
    if df['sentiment'][i]=='love':
        sentimentint[i]='6'
        continue
    if df['sentiment'][i]=='fun':
        sentimentint[i]='7'
        continue
    if df['sentiment'][i]=='happiness':
        sentimentint[i]='8'
        continue
    if df['sentiment'][i]=='hate':
        sentimentint[i]='9'
        continue
    if df['sentiment'][i]=='boredom':
        sentimentint[i]='10'
        continue
    if df['sentiment'][i]=='anger':
        sentimentint[i]='11'
        continue
    if df['sentiment'][i]=='relief':
        sentimentint[i]='12'
        continue
    sentimentint[i]='0'
        
df['sentimentint'] = sentimentint    



In [386]:
df.head(10)


tweet_id   sentiment         author  \
0  1956967341       empty     xoshayzers   
1  1956967666     sadness      wannamama   
2  1956967696     sadness      coolfunky   
3  1956967789  enthusiasm    czareaquino   
4  1956968416     neutral      xkilljoyx   
5  1956968477       worry  xxxPEACHESxxx   
6  1956968487     sadness       ShansBee   
7  1956968636       worry       mcsleazy   
8  1956969035     sadness    nic0lepaula   
9  1956969172     sadness     Ingenue_Em   

                                             content sentimentint  
0  @tiffanylue i know  i was listenin to bad habi...            0  
1  Layin n bed with a headache  ughhhh...waitin o...            1  
2                Funeral ceremony...gloomy friday...            1  
3               wants to hang out with friends SOON!            2  
4  @dannycastillo We want to trade with someone w...            3  
5  Re-pinging @ghostridah14: why didn't you go to...            4  
6  I should be sleep, but im not! thinking about ...            1  
7               Hmmm. http://www.djhero.com/ is down            4  
8            @charviray Charlene my love. I miss you            1  
9         @kelcouch I'm sorry  at least it's Friday?            1

##### preprocesamiento de texto df["content"]

In [387]:
from nltk.corpus import stopwords
#función remove stopWords
cachedStopWords = stopwords.words("english")
def stopWords(text):
    text = ' '.join([word for word in text.split() if word not in cachedStopWords])
    return text

from nltk.stem.snowball import SnowballStemmer
#función apply stemmer
stemmer = SnowballStemmer("english")
def stemmerWords(text):
    text = ' '.join([stemmer.stem(word) for word in text.split()])
    return text



# transformamos los datos con la función lower()
df["content"]=df["content"].str.lower()

#idea... agregar una columna extra al final que cuente la cantidad de etiquetas @etiqueta
# eliminamos todas las @etiquetas de los usuarios
df["content"] = df["content"].apply(lambda x: " ".join(x for x in x.split() if not x.startswith('@')))

# quitamos todos los carácteres que no son ni letras ni números
df["content"]=df["content"].replace('[^a-z0-9]', ' ', regex = True)

# df["content"]=df["content"].replace(r'(^.*mm.*$)', 'mmmm')
df["content"] = df["content"].replace('[a-zA-Z]*m{2}[a-zA-Z]*', 'mmm', regex = True)

# df["content"]=df["content"].replace(r'(^.*mm.*$)', 'mmmm')
df["content"] = df["content"].replace('[a-zA-Z]*ugh[a-zA-Z]*', 'ugh', regex = True)


#quitando stopwords
# df["content"] = df["content"].apply(lambda x: stopWords(x))

#Stemmer words 
df["content"] = df["content"].apply(lambda x: stemmerWords(x))






Aquí se trabaja la letra B

In [388]:
msk = np.random.rand(len(df)) < 0.9
df_train = df[msk]
df_test = df[~msk]
print "df shape= " +str(df.shape)
print "df train shape= " +str(df_train.shape)
print "df test shape= " +str(df_test.shape)
df.head(20)

df shape= (40000, 5)
df train shape= (35970, 5)
df test shape= (4030, 5)


tweet_id   sentiment           author  \
0   1956967341       empty       xoshayzers   
1   1956967666     sadness        wannamama   
2   1956967696     sadness        coolfunky   
3   1956967789  enthusiasm      czareaquino   
4   1956968416     neutral        xkilljoyx   
5   1956968477       worry    xxxPEACHESxxx   
6   1956968487     sadness         ShansBee   
7   1956968636       worry         mcsleazy   
8   1956969035     sadness      nic0lepaula   
9   1956969172     sadness       Ingenue_Em   
10  1956969456     neutral       feinyheiny   
11  1956969531       worry     dudeitsmanda   
12  1956970047     sadness         Danied32   
13  1956970424     sadness          Samm_xo   
14  1956970860    surprise     okiepeanut93   
15  1956971077     sadness           Sim_34   
16  1956971170        love     poppygallico   
17  1956971206     sadness  brokenangel1982   
18  1956971473       worry            LCJ82   
19  1956971586     sadness          cleepow   

                                              content sentimentint  
0   i know i was listenin to bad habit earlier and...            0  
1   layin n bed with a headach ugh waitin on your ...            1  
2                        funer ceremoni gloomi friday            1  
3                   want to hang out with friend soon            2  
4   we want to trade with someon who has houston t...            3  
5   re ping whi didn t you go to prom bc my bf did...            4  
6   i should be sleep but im not think about an ol...            1  
7                     mmm http www djhero com is down            4  
8                          charlen my love i miss you            1  
9                      i m sorri at least it s friday            1  
10                                   cant fall asleep            3  
11                                choke on her retain            4  
12  ugh i have to beat this stupid song to get to ...            1  
13  if u watch the hill in london u will realis wh...            1  
14                                       got the news            5  
15           the storm is here and the electr is gone            1  
16                                               agre            6  
17  so sleepi again and it s not even that late i ...            1  
18  ladi gaga tweet about not be impress by her vi...            4  
19  how are you convinc that i have alway want you...            1

In [389]:
df_train["content"][0:10]


0    i know i was listenin to bad habit earlier and...
1    layin n bed with a headach ugh waitin on your ...
2                         funer ceremoni gloomi friday
3                    want to hang out with friend soon
4    we want to trade with someon who has houston t...
5    re ping whi didn t you go to prom bc my bf did...
6    i should be sleep but im not think about an ol...
7                      mmm http www djhero com is down
8                           charlen my love i miss you
9                       i m sorri at least it s friday
Name: content, dtype: object

In [390]:
df_train["sentimentint"][0:10]

0    0
1    1
2    1
3    2
4    3
5    4
6    1
7    4
8    1
9    1
Name: sentimentint, dtype: object

In [391]:
from sklearn.feature_extraction.text import CountVectorizer
from scipy.sparse import hstack
# vectorizer = CountVectorizer()
vectorizer = CountVectorizer(ngram_range=(1, 2),min_df=2,binary='False')
vectorizer.fit(df_train["content"])

#obtenemos las matrices de features utilizando la función vector anterior
X_tfidf = vectorizer.transform(df_train["content"])
X_vfidf = vectorizer.transform(df_test["content"]) 

X = hstack([X_tfidf])
Xval = hstack([X_vfidf])

y = df_train["sentimentint"]
yval = df_test["sentimentint"]



>d) Entrene y compare al menos 4 de los diferentes clasificadores vistos en clases para clasificación (por ejemplo: Navie Bayes, Multinomial Naive Bayes, LDA, QDA, Regresión logı́stica y Perceptrón). Recuerde que algunos son extendidos por defecto a múltiples clases para detectar emociones en cada tweet, sin embargo, otros deben ser extentidos a través de otras técnicas, tal como One vs One y One vs All/Rest. Muestre tabla o gráfico resumen.

In [392]:
#utilizando Multinomial naive bayes
from sklearn.naive_bayes import MultinomialNB
modelMNB = MultinomialNB()
modelMNB.fit(X, y)
print modelMNB.score(X, y)
print modelMNB.score(Xval, yval)

from sklearn.multiclass import OneVsRestClassifier
from sklearn.multiclass import OneVsOneClassifier

classifMNB = OneVsRestClassifier(MultinomialNB())
classifMNB.fit(X, y)
print classifMNB.score(X,y)
print classifMNB.score(Xval,yval)
#SE APRECIAN MEJORAS TRAS APLICAR OneVSRestClassifier

0.6261050875729774
0.33523573200992557
0.6552682791214901
0.3389578163771712


In [393]:
from sklearn.linear_model import Perceptron
Cs = 100000
modelPerceptron = Perceptron(penalty='l2', alpha=1.0/Cs, max_iter=200)
modelPerceptron.fit(X,y)

print modelPerceptron.score(X,y)
print modelPerceptron.score(Xval,yval)

from sklearn.multiclass import OneVsRestClassifier
from sklearn.multiclass import OneVsOneClassifier
#example
classifPERCEP = OneVsOneClassifier(Perceptron(penalty='l2', alpha=1.0/Cs, max_iter=200))
classifPERCEP.fit(X, y)
print classifPERCEP.score(X,y)
print classifPERCEP.score(Xval,yval)

0.5782040589380039
0.23225806451612904
0.808868501529052
0.29851116625310176


In [394]:
from sklearn.linear_model import LogisticRegression
Cs = 100
modelLogistic = LogisticRegression(penalty='l2', C=Cs,max_iter=200)
modelLogistic.fit(X,y)   
print modelLogistic.score(X,y)
print modelLogistic.score(Xval,yval)

classifLOGIS = OneVsRestClassifier(LogisticRegression(penalty='l2', C=Cs,max_iter=200))
classifLOGIS.fit(X, y)
print classifLOGIS.score(X,y)
print classifLOGIS.score(Xval,yval)

0.981845982763414
0.2826302729528536
0.9818181818181818
0.2826302729528536


In [ ]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
modelQDA = QuadraticDiscriminantAnalysis()
modelQDA.fit(asdf, y)
print modelQDA.score(X, y)
print modelQDA.score(Xval, yval)

from sklearn.multiclass import OneVsRestClassifier
from sklearn.multiclass import OneVsOneClassifier

classifQDA= OneVsRestClassifier(GaussianNB())
classifQDA.fit(X, y)
print classifQDA.score(X,y)

In [ ]:
No logramos hacer funcionar el método QDA debido a problemas con el lenguaje.

>e) Utilice la técnica de ECOC (Error-Correcting Output-Code) para extender a multiclases algunos de los clasificadores utilizados en d). Comente lo que hace la técnica y los resultados observados.

In [395]:
from sklearn.multiclass import OutputCodeClassifier
clf = OutputCodeClassifier(Perceptron(penalty='l2', alpha=1.0/100000, max_iter=200),code_size=2, random_state=0)
clf.fit(X, y)
print clf.score(X,y)
print clf.score(Xval,yval)


TypeError: A sparse matrix was passed, but dense data is required. Use X.toarray() to convert to a dense numpy array.

No logramos hacer funcionar ECOC debido a problemas con el lenguaje.

>f) Evalúe la métrica de accuracy sobre el conjunto de pruebas del mejor clasificador encontrado.
Recuerde que puede acudir a otras métricas para tener otras visiones de lo que está haciendo el modelo de aprendizaje

In [396]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB(alpha=1.2)
model.fit(X, y)

from sklearn.metrics import accuracy_score
y_pred = model.predict(Xval)
y_true = yval
accuracy_score(y_true, y_pred)

0.33548387096774196

>g) Intente mejorar su resultado considerablemente a través de alguna mejora novedosa. Se espera que supere el 35% de accuracy.

El máximo obtenido de accuracy fue de 0.335 
No se pudo mejorar.
Para obtener este accuracy se removieron los tags del tipo @nombre de los tweets